File: omni/extensions/runtime/tests/test.unit/physics/TestSimulatorStageUpdate.cpp
// SPDX-FileCopyrightText: Copyright (c) 2020-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause
//

#include "UsdPCH.h"

#include "../common/TestHelpers.h"

#include "PhysicsTools.h"

#include <omni/physics/simulation/IPhysics.h>
#include <omni/physics/simulation/IPhysicsStageUpdate.h>
#include <carb/events/IEvents.h>
#include <carb/events/EventsUtils.h>

using namespace omni::physics;

// Mock stage update implementation
class MockStageUpdate
{
public:
    enum class TimelineState
    {
        Stopped,
        Paused,
        Playing
    };

    MockStageUpdate()
        : m_attachedStageId(0)
        , m_isAttached(false)
        , m_isPhysicsLoaded(false)
        , m_timelineState(TimelineState::Stopped)
        , m_resetSimulationCalled(false)
        , m_startSimulationCalled(false)
    {
        // Initialize function pointers
        onAttach = [this](long int stageId) {
            m_attachedStageId = stageId;
            m_isAttached = true;
        };

        onDetach = [this]() {
            m_attachedStageId = 0;
            m_isAttached = false;
            m_isPhysicsLoaded = false;
            m_timelineState = TimelineState::Stopped;
        };

        onUpdate = [this](float currentTime, float elapsedSecs, bool enableUpdate) {
            if (m_isAttached && enableUpdate) {
                // Mock update implementation
            }
        };

        onResume = [this](float currentTime) {
            if (m_isAttached) {
                m_timelineState = TimelineState::Playing;
            }
        };

        onPause = [this]() {
            if (m_isAttached) {
                m_timelineState = TimelineState::Paused;
            }
        };

        onReset = [this]() {
            if (m_isAttached) {
                m_timelineState = TimelineState::Stopped;
            }
        };

        handleRaycast = [this](const float* orig, const float* dir, bool input) {
            if (m_isAttached) {
                // Mock raycast implementation
            }
        };

        forceLoadPhysicsFromUSD = [this]() {
            if (m_isAttached) {
                m_isPhysicsLoaded = true;
            }
        };

        releasePhysicsObjects = [this]() {
            if (m_isAttached) {
                m_isPhysicsLoaded = false;
            }
        };

        resetSimulation = [this]() {
            if (m_isAttached) {
                m_isPhysicsLoaded = false;
                m_timelineState = TimelineState::Stopped;
                m_resetSimulationCalled = true;
            }
        };

        startSimulation = [this]() {
            if (m_isAttached) {
                m_startSimulationCalled = true;
            }
        };
    }

    ~MockStageUpdate()
    {
        // Cleanup
    }

    // Function pointers that match StageUpdateFns structure
    OnAttachFn onAttach;
    OnDetachFn onDetach;
    OnUpdateFn onUpdate;
    OnResumeFn onResume;
    OnPauseFn onPause;
    OnResetFn onReset;
    HandleRaycastFn handleRaycast;
    ForceLoadPhysicsFromUSDFn forceLoadPhysicsFromUSD;
    ReleasePhysicsObjectsFn releasePhysicsObjects;
    ResetSimulationFn resetSimulation;
    StartSimulationFn startSimulation;

    // Get the function pointers structure
    StageUpdateFns getStageUpdateFns() const
    {
        StageUpdateFns fns;
        fns.onAttach = onAttach;
        fns.onDetach = onDetach;
        fns.onUpdate = onUpdate;
        fns.onResume = onResume;
        fns.onPause = onPause;
        fns.onReset = onReset;
        fns.handleRaycast = handleRaycast;
        fns.forceLoadPhysicsFromUSD = forceLoadPhysicsFromUSD;
        fns.releasePhysicsObjects = releasePhysicsObjects;
        fns.resetSimulation = resetSimulation;
        fns.startSimulation = startSimulation;
        return fns;
    }

    TimelineState getTimelineState() const { return m_timelineState; }
    bool wasResetSimulationCalled() const { return m_resetSimulationCalled; }
    void clearResetSimulationFlag() { m_resetSimulationCalled = false; }
    bool wasStartSimulationCalled() const { return m_startSimulationCalled; }
    void clearStartSimulationFlag() { m_startSimulationCalled = false; }

public:
    // Internal state
    long m_attachedStageId;
    bool m_isAttached;
    bool m_isPhysicsLoaded;
    TimelineState m_timelineState;
    bool m_resetSimulationCalled;
    bool m_startSimulationCalled;
};

//-----------------------------------------------------------------------------
// Stage Update Tests
TEST_CASE("Simulator Stage Update Tests",
    "[omniphysics]"
    "[component=OmniPhysics][owner=aborovicka][priority=mandatory]")
{
    ScopedOmniPhysicsActivation scopedOmniPhysicsActivation;

    PhysicsTest& physicsTests = *PhysicsTest::getPhysicsTests();
    IPhysics* physics = physicsTests.getApp()->getFramework()->acquireInterface<IPhysics>();
    REQUIRE(physics);

    IPhysicsStageUpdate* stageUpdate = physicsTests.getApp()->getFramework()->acquireInterface<IPhysicsStageUpdate>();
    REQUIRE(stageUpdate);

    // Create a mock stage update
    MockStageUpdate mockStageUpdate;
    StageUpdateFns fns = mockStageUpdate.getStageUpdateFns();

    Simulation sim;
    sim.stageUpdateFns = fns;

    SimulationId simId = physics->registerSimulation(sim, "MockupSimulator");
    REQUIRE(simId != kInvalidSimulationId);

    SECTION("Stage attachment")
    {
        stageUpdate->onAttach(123);
        REQUIRE(mockStageUpdate.m_attachedStageId == 123);
        REQUIRE(mockStageUpdate.m_isAttached);
        
        stageUpdate->onDetach();
        REQUIRE(mockStageUpdate.m_attachedStageId == 0);
        REQUIRE_FALSE(mockStageUpdate.m_isAttached);
    }

    SECTION("Stage update")
    {
        stageUpdate->onAttach(123);
        
        // Test update with physics enabled
        stageUpdate->onUpdate(1.0f, 0.016f, true);
        REQUIRE(mockStageUpdate.m_isAttached);
        
        // Test update with physics disabled
        stageUpdate->onUpdate(1.0f, 0.016f, false);
        REQUIRE(mockStageUpdate.m_isAttached);
    }

    SECTION("Timeline control")
    {
        stageUpdate->onAttach(123);
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Stopped);
        
        // Test resume
        stageUpdate->onResume(1.0f);
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Playing);
        
        // Test pause
        stageUpdate->onPause();
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Paused);
        
        // Test reset
        stageUpdate->onReset();
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Stopped);

        // Test detach resets timeline state
        stageUpdate->onDetach();
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Stopped);
    }

    SECTION("Physics loading")
    {
        stageUpdate->onAttach(123);
        
        // Test force load
        stageUpdate->forceLoadPhysicsFromUSD();
        REQUIRE(mockStageUpdate.m_isPhysicsLoaded);
        
        // Test release
        stageUpdate->releasePhysicsObjects();
        REQUIRE_FALSE(mockStageUpdate.m_isPhysicsLoaded);
    }

    SECTION("Raycast handling")
    {
        stageUpdate->onAttach(123);
        
        const float orig[3] = { 0.0f, 0.0f, 0.0f };
        const float dir[3] = { 0.0f, 0.0f, 1.0f };
        
        // Test raycast with input
        stageUpdate->handleRaycast(orig, dir, true);
        REQUIRE(mockStageUpdate.m_isAttached);
        
        // Test raycast without input
        stageUpdate->handleRaycast(orig, dir, false);
        REQUIRE(mockStageUpdate.m_isAttached);
    }

    SECTION("Reset simulation")
    {
        stageUpdate->onAttach(123);
        REQUIRE(mockStageUpdate.m_isAttached);
        REQUIRE_FALSE(mockStageUpdate.wasResetSimulationCalled());
        
        // Load physics first
        stageUpdate->forceLoadPhysicsFromUSD();
        REQUIRE(mockStageUpdate.m_isPhysicsLoaded);
        
        // Set timeline to playing
        stageUpdate->onResume(1.0f);
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Playing);
        
        // Test resetSimulation
        stageUpdate->resetSimulation();
        
        // Verify simulation was reset
        REQUIRE(mockStageUpdate.wasResetSimulationCalled());
        REQUIRE_FALSE(mockStageUpdate.m_isPhysicsLoaded);  // Physics should be released
        REQUIRE(mockStageUpdate.getTimelineState() == MockStageUpdate::TimelineState::Stopped);  // Timeline should be stopped
        REQUIRE(mockStageUpdate.m_isAttached);  // Stage should remain attached
        
        // Test that we can clear the flag and call again
        mockStageUpdate.clearResetSimulationFlag();
        REQUIRE_FALSE(mockStageUpdate.wasResetSimulationCalled());
        
        stageUpdate->resetSimulation();
        REQUIRE(mockStageUpdate.wasResetSimulationCalled());
    }

    SECTION("Start simulation")
    {
        stageUpdate->onAttach(123);
        REQUIRE(mockStageUpdate.m_isAttached);
        REQUIRE_FALSE(mockStageUpdate.wasStartSimulationCalled());
        
        // Test startSimulation
        stageUpdate->startSimulation();
        
        // Verify simulation was started
        REQUIRE(mockStageUpdate.wasStartSimulationCalled());
        REQUIRE(mockStageUpdate.m_isAttached);  // Stage should remain attached
        
        // Test that we can clear the flag and call again
        mockStageUpdate.clearStartSimulationFlag();
        REQUIRE_FALSE(mockStageUpdate.wasStartSimulationCalled());
        
        stageUpdate->startSimulation();
        REQUIRE(mockStageUpdate.wasStartSimulationCalled());
        
        stageUpdate->onDetach();
    }

    physics->unregisterSimulation(simId);
}

//-----------------------------------------------------------------------------
// Simulation Event Stream Tests
TEST_CASE("Simulation Event Stream Tests",
    "[omniphysics]"
    "[component=OmniPhysics][owner=aborovicka][priority=mandatory]")
{
    ScopedOmniPhysicsActivation scopedOmniPhysicsActivation;

    PhysicsTest& physicsTests = *PhysicsTest::getPhysicsTests();
    IPhysics* physics = physicsTests.getApp()->getFramework()->acquireInterface<IPhysics>();
    REQUIRE(physics);

    IPhysicsStageUpdate* stageUpdate = physicsTests.getApp()->getFramework()->acquireInterface<IPhysicsStageUpdate>();
    REQUIRE(stageUpdate);

    // Create a mock stage update
    MockStageUpdate mockStageUpdate;
    StageUpdateFns fns = mockStageUpdate.getStageUpdateFns();

    Simulation sim;
    sim.stageUpdateFns = fns;

    SimulationId simId = physics->registerSimulation(sim, "MockupSimulatorEvents");
    REQUIRE(simId != kInvalidSimulationId);

    SECTION("Event stream exists")
    {
        carb::events::IEventStreamPtr eventStream = stageUpdate->getSimulationEventStream();
        REQUIRE(eventStream != nullptr);
    }

    SECTION("Resume event is sent")
    {
        carb::events::IEventStreamPtr eventStream = stageUpdate->getSimulationEventStream();
        REQUIRE(eventStream != nullptr);

        // Track received events
        std::vector<SimulationEvent> receivedEvents;
        auto subscription = carb::events::createSubscriptionToPop(
            eventStream.get(),
            [&receivedEvents](carb::events::IEvent* event) {
                receivedEvents.push_back(static_cast<SimulationEvent>(event->type));
            }
        );

        stageUpdate->onAttach(123);
        stageUpdate->onResume(1.0f);

        // Process events
        eventStream->pump();

        REQUIRE(receivedEvents.size() == 1);
        REQUIRE(receivedEvents[0] == SimulationEvent::eResumed);

        stageUpdate->onDetach();
    }

    SECTION("Pause event is sent")
    {
        carb::events::IEventStreamPtr eventStream = stageUpdate->getSimulationEventStream();
        REQUIRE(eventStream != nullptr);

        // Track received events
        std::vector<SimulationEvent> receivedEvents;
        auto subscription = carb::events::createSubscriptionToPop(
            eventStream.get(),
            [&receivedEvents](carb::events::IEvent* event) {
                receivedEvents.push_back(static_cast<SimulationEvent>(event->type));
            }
        );

        stageUpdate->onAttach(123);
        stageUpdate->onResume(1.0f);
        eventStream->pump();
        receivedEvents.clear();

        stageUpdate->onPause();

        // Process events
        eventStream->pump();

        REQUIRE(receivedEvents.size() == 1);
        REQUIRE(receivedEvents[0] == SimulationEvent::ePaused);

        stageUpdate->onDetach();
    }

    SECTION("Stopped event is sent on reset")
    {
        carb::events::IEventStreamPtr eventStream = stageUpdate->getSimulationEventStream();
        REQUIRE(eventStream != nullptr);

        // Track received events
        std::vector<SimulationEvent> receivedEvents;
        auto subscription = carb::events::createSubscriptionToPop(
            eventStream.get(),
            [&receivedEvents](carb::events::IEvent* event) {
                receivedEvents.push_back(static_cast<SimulationEvent>(event->type));
            }
        );

        stageUpdate->onAttach(123);
        stageUpdate->onResume(1.0f);
        eventStream->pump();
        receivedEvents.clear();

        stageUpdate->onReset();

        // Process events
        eventStream->pump();

        REQUIRE(receivedEvents.size() == 1);
        REQUIRE(receivedEvents[0] == SimulationEvent::eStopped);

        stageUpdate->onDetach();
    }

    SECTION("Full lifecycle events")
    {
        carb::events::IEventStreamPtr eventStream = stageUpdate->getSimulationEventStream();
        REQUIRE(eventStream != nullptr);

        // Track received events
        std::vector<SimulationEvent> receivedEvents;
        auto subscription = carb::events::createSubscriptionToPop(
            eventStream.get(),
            [&receivedEvents](carb::events::IEvent* event) {
                receivedEvents.push_back(static_cast<SimulationEvent>(event->type));
            }
        );

        stageUpdate->onAttach(123);

        // Resume -> Pause -> Resume -> Stop
        stageUpdate->onResume(1.0f);
        eventStream->pump();
        REQUIRE(receivedEvents.back() == SimulationEvent::eResumed);

        stageUpdate->onPause();
        eventStream->pump();
        REQUIRE(receivedEvents.back() == SimulationEvent::ePaused);

        stageUpdate->onResume(2.0f);
        eventStream->pump();
        REQUIRE(receivedEvents.back() == SimulationEvent::eResumed);

        stageUpdate->onReset();
        eventStream->pump();
        REQUIRE(receivedEvents.back() == SimulationEvent::eStopped);

        REQUIRE(receivedEvents.size() == 4);

        stageUpdate->onDetach();
    }

    physics->unregisterSimulation(simId);
} 

File: omni/umbrella/tests/python/testSimulatorStageUpdate.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#

import _carb_setup  # noqa: F401 - initializes Carbonite framework

import carb
import unittest
from _physics import (
    Simulation,
    acquire_physics_interface,
    acquire_physics_stage_update_interface,
    k_invalid_simulation_id,
    SimulationEvent,
)


class MockPhysicsStageUpdate:
    class TimelineState:
        STOPPED = 0
        PAUSED = 1
        PLAYING = 2

    def __init__(self):
        self.attached_stage_id = None
        self.is_physics_enabled = False
        self.is_paused = False
        self.update_count = 0
        self.raycast_hits = []
        self.force_load_called = False
        self.release_objects_called = False
        self.is_physics_loaded = False
        self.timeline_state = self.TimelineState.STOPPED
        self.reset_simulation_called = False
        self.start_simulation_called = False

    def on_attach(self, stage_id):
        self.attached_stage_id = stage_id
        return True

    def on_detach(self):
        self.attached_stage_id = None
        self.is_physics_loaded = False
        self.timeline_state = self.TimelineState.STOPPED
        return True

    def on_update(self, time, delta_time, physics_enabled):
        self.is_physics_enabled = physics_enabled
        self.update_count += 1
        return True

    def on_resume(self, time):
        if self.attached_stage_id is not None:
            self.timeline_state = self.TimelineState.PLAYING
        self.is_paused = False
        return True

    def on_pause(self):
        if self.attached_stage_id is not None:
            self.timeline_state = self.TimelineState.PAUSED
        self.is_paused = True
        return True

    def on_reset(self):
        if self.attached_stage_id is not None:
            self.timeline_state = self.TimelineState.STOPPED
        self.update_count = 0
        self.is_paused = False
        return True

    def force_load_physics_from_usd(self):
        if self.attached_stage_id is not None:
            self.is_physics_loaded = True
        self.force_load_called = True
        return True

    def release_physics_objects(self):
        if self.attached_stage_id is not None:
            self.is_physics_loaded = False
        self.release_objects_called = True
        return True

    def reset_simulation(self):
        if self.attached_stage_id is not None:
            self.is_physics_loaded = False
            self.timeline_state = self.TimelineState.STOPPED
            self.reset_simulation_called = True
        return True

    def start_simulation(self):
        if self.attached_stage_id is not None:
            self.start_simulation_called = True
        return True

    def handle_raycast(self, origin, direction, has_input):
        if has_input:
            self.raycast_hits.append((origin, direction))
        return True


class TestSimulateStageUpdate(unittest.TestCase):
    def setUp(self):
        self.physics = acquire_physics_interface()
        self.physics_stage_update = acquire_physics_stage_update_interface()
        self.mock = MockPhysicsStageUpdate()

        self.simulation = Simulation()
        self.simulation.stage_update_fns.on_attach = self.mock.on_attach
        self.simulation.stage_update_fns.on_detach = self.mock.on_detach
        self.simulation.stage_update_fns.on_update = self.mock.on_update
        self.simulation.stage_update_fns.on_resume = self.mock.on_resume
        self.simulation.stage_update_fns.on_pause = self.mock.on_pause
        self.simulation.stage_update_fns.on_reset = self.mock.on_reset
        self.simulation.stage_update_fns.force_load_physics_from_usd = self.mock.force_load_physics_from_usd
        self.simulation.stage_update_fns.release_physics_objects = self.mock.release_physics_objects
        self.simulation.stage_update_fns.handle_raycast = self.mock.handle_raycast
        self.simulation.stage_update_fns.reset_simulation = self.mock.reset_simulation
        self.simulation.stage_update_fns.start_simulation = self.mock.start_simulation

        self.simulation_id = self.physics.register_simulation(self.simulation, "MockStageUpdateSim")
        self.assertNotEqual(self.simulation_id, k_invalid_simulation_id)

    def tearDown(self):
        if self.simulation_id != k_invalid_simulation_id:
            self.physics.unregister_simulation(self.simulation_id)

    def test_stage_attachment(self):
        self.physics_stage_update.on_attach(123)
        self.assertEqual(self.mock.attached_stage_id, 123)
        self.physics_stage_update.on_detach()
        self.assertIsNone(self.mock.attached_stage_id)

    def test_stage_update(self):
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.on_update(1.0, 0.016, True)
        self.assertTrue(self.mock.is_physics_enabled)
        self.assertEqual(self.mock.update_count, 1)
        self.physics_stage_update.on_update(1.0, 0.016, False)
        self.assertFalse(self.mock.is_physics_enabled)
        self.physics_stage_update.on_detach()

    def test_timeline_control(self):
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.on_resume(1.0)
        self.assertFalse(self.mock.is_paused)
        self.physics_stage_update.on_pause()
        self.assertTrue(self.mock.is_paused)
        self.physics_stage_update.on_reset()
        self.assertEqual(self.mock.update_count, 0)
        self.assertFalse(self.mock.is_paused)
        self.physics_stage_update.on_detach()

    def test_physics_loading(self):
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.force_load_physics_from_usd()
        self.assertTrue(self.mock.is_physics_loaded)
        self.physics_stage_update.release_physics_objects()
        self.assertFalse(self.mock.is_physics_loaded)
        self.physics_stage_update.on_detach()

    def test_raycast_handling(self):
        self.physics_stage_update.on_attach(123)
        origin = carb.Float3(0.0, 0.0, 0.0)
        direction = carb.Float3(0.0, 0.0, 1.0)
        self.physics_stage_update.handle_raycast(origin, direction, True)
        self.assertEqual(len(self.mock.raycast_hits), 1)
        self.physics_stage_update.handle_raycast(origin, direction, False)
        self.assertEqual(len(self.mock.raycast_hits), 1)
        self.physics_stage_update.on_detach()

    def test_reset_simulation(self):
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.force_load_physics_from_usd()
        self.physics_stage_update.on_resume(1.0)
        self.physics_stage_update.reset_simulation()
        self.assertTrue(self.mock.reset_simulation_called)
        self.assertFalse(self.mock.is_physics_loaded)
        self.assertEqual(self.mock.timeline_state, MockPhysicsStageUpdate.TimelineState.STOPPED)
        self.physics_stage_update.on_detach()

    def test_start_simulation(self):
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.start_simulation()
        self.assertTrue(self.mock.start_simulation_called)
        self.physics_stage_update.on_detach()

    def test_simulation_event_stream_exists(self):
        event_stream = self.physics_stage_update.get_simulation_event_stream()
        self.assertIsNotNone(event_stream)

    def test_simulation_event_resumed(self):
        event_stream = self.physics_stage_update.get_simulation_event_stream()
        received_events = []

        def on_event(event):
            received_events.append(event.type)

        subscription = event_stream.create_subscription_to_pop(on_event)
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.on_resume(1.0)
        event_stream.pump()

        self.assertEqual(len(received_events), 1)
        self.assertEqual(received_events[0], SimulationEvent.RESUMED)
        self.physics_stage_update.on_detach()

    def test_simulation_event_paused(self):
        event_stream = self.physics_stage_update.get_simulation_event_stream()
        received_events = []

        def on_event(event):
            received_events.append(event.type)

        subscription = event_stream.create_subscription_to_pop(on_event)
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.on_resume(1.0)
        event_stream.pump()
        received_events.clear()

        self.physics_stage_update.on_pause()
        event_stream.pump()
        self.assertEqual(len(received_events), 1)
        self.assertEqual(received_events[0], SimulationEvent.PAUSED)
        self.physics_stage_update.on_detach()

    def test_simulation_event_stopped(self):
        event_stream = self.physics_stage_update.get_simulation_event_stream()
        received_events = []

        def on_event(event):
            received_events.append(event.type)

        subscription = event_stream.create_subscription_to_pop(on_event)
        self.physics_stage_update.on_attach(123)
        self.physics_stage_update.on_resume(1.0)
        event_stream.pump()
        received_events.clear()

        self.physics_stage_update.on_reset()
        event_stream.pump()
        self.assertEqual(len(received_events), 1)
        self.assertEqual(received_events[0], SimulationEvent.STOPPED)
        self.physics_stage_update.on_detach()

    def test_simulation_event_full_lifecycle(self):
        event_stream = self.physics_stage_update.get_simulation_event_stream()
        received_events = []

        def on_event(event):
            received_events.append(event.type)

        subscription = event_stream.create_subscription_to_pop(on_event)
        self.physics_stage_update.on_attach(123)

        self.physics_stage_update.on_resume(1.0)
        event_stream.pump()
        self.assertEqual(received_events[-1], SimulationEvent.RESUMED)

        self.physics_stage_update.on_pause()
        event_stream.pump()
        self.assertEqual(received_events[-1], SimulationEvent.PAUSED)

        self.physics_stage_update.on_resume(2.0)
        event_stream.pump()
        self.assertEqual(received_events[-1], SimulationEvent.RESUMED)

        self.physics_stage_update.on_reset()
        event_stream.pump()
        self.assertEqual(received_events[-1], SimulationEvent.STOPPED)

        self.assertEqual(len(received_events), 4)
        self.physics_stage_update.on_detach()


if __name__ == "__main__":
    unittest.main()
